File: examples_python/example_basic.py
#!/usr/bin/env python3
"""
libzupt - Basic Encryption/Decryption Example
Shows how to use the library for basic memory encryption/decryption
"""

import sys
sys.path.insert(0, __import__('os').path.dirname(__file__))

import zupt


def main():
    print("=" * 60)
    print("libzupt - Basic Encryption/Decryption Example")
    print("=" * 60)
    print()

    # Generate key pair
    print("1. Generating key pair...")
    keygen = zupt.KeyGenerator()
    keypair = keygen.generate_keypair()
    print(f"   Public key size: {len(keypair.public_key)} bytes")
    print(f"   Secret key size: {len(keypair.secret_key)} bytes")
    print()

    # Create encryptor and decryptor
    encryptor = zupt.Encryptor(keypair.public_key)
    decryptor = zupt.Decryptor(keypair.secret_key)

    # Encrypt some data
    message = b"Hello, Post-Quantum World! This is a secret message."
    print(f"2. Encrypting message: {message.decode()}")
    ciphertext, enc_header = encryptor.encrypt(message)
    print(f"   Ciphertext size: {len(ciphertext)} bytes")
    print(f"   Header size: {len(enc_header)} bytes")
    print()

    # Decrypt
    print("3. Decrypting...")
    decrypted = decryptor.decrypt(ciphertext, list(enc_header))
    print(f"   Decrypted: {decrypted.decode()}")
    print()

    # Verify
    assert decrypted == message, "Decryption failed!"
    print("4. Verification: SUCCESS - Decrypted message matches original")
    print()

    # Demonstrate error with wrong key
    print("5. Testing with wrong key...")
    keygen2 = zupt.KeyGenerator()
    keypair2 = keygen2.generate_keypair()
    decryptor_wrong = zupt.Decryptor(keypair2.secret_key)

    try:
        decryptor_wrong.decrypt(ciphertext, list(enc_header))
        print("   ERROR: Should have failed!")
    except RuntimeError as e:
        print(f"   Correctly rejected with error: {e}")
    print()

    print("=" * 60)
    print("All examples passed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: examples_dotnet/ExampleBasic.cs
using LibZupt;
using System.Text;

namespace ZuptExample;

internal static class ExampleBasic
{
    public static void Run()
    {
        Console.WriteLine(new string('=', 60));
        Console.WriteLine("libzupt - Basic Encryption/Decryption Example");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine();

        Console.WriteLine("1. Generating key pair...");
        var keygen = new KeyGenerator();
        var keypair = keygen.GenerateKeyPair();
        Console.WriteLine($"   Public key size: {keypair.PublicKey.Length} bytes");
        Console.WriteLine($"   Secret key size: {keypair.SecretKey.Length} bytes");
        Console.WriteLine();

        var encryptor = new Encryptor(keypair.PublicKey);
        var decryptor = new Decryptor(keypair.SecretKey);

        byte[] message = Encoding.UTF8.GetBytes("Hello, Post-Quantum World! This is a secret message.");
        Console.WriteLine($"2. Encrypting message: {Encoding.UTF8.GetString(message)}");
        var (ciphertext, encHeader) = encryptor.EncryptMemory(message);
        Console.WriteLine($"   Ciphertext size: {ciphertext.Length} bytes");
        Console.WriteLine($"   Header size: {encHeader.Length} bytes");
        Console.WriteLine();

        Console.WriteLine("3. Decrypting...");
        byte[] decrypted = decryptor.DecryptMemory(ciphertext, encHeader);
        Console.WriteLine($"   Decrypted: {Encoding.UTF8.GetString(decrypted)}");
        Console.WriteLine();

        if (!decrypted.SequenceEqual(message))
            throw new ZuptException("Decryption failed!");
        Console.WriteLine("4. Verification: SUCCESS - Decrypted message matches original");
        Console.WriteLine();

        Console.WriteLine("5. Testing with wrong key...");
        var keygen2 = new KeyGenerator();
        var keypair2 = keygen2.GenerateKeyPair();
        var decryptorWrong = new Decryptor(keypair2.SecretKey);

        try
        {
            decryptorWrong.DecryptMemory(ciphertext, encHeader);
            Console.WriteLine("   ERROR: Should have failed!");
        }
        catch (ZuptException e)
        {
            Console.WriteLine($"   Correctly rejected with error: {e.Message}");
        }
        Console.WriteLine();

        Console.WriteLine(new string('=', 60));
        Console.WriteLine("All examples passed!");
        Console.WriteLine(new string('=', 60));
        Console.WriteLine();
    }
}